### DISCUSSION TITLE
Container generated as a SystemJS library loads its exposed modules from host url

### DISCUSSION THREAD

**Discussion by fbessou** [webpack/webpack#13522] [General]
# Bug report

**What is the current behavior?**
A container library generated using the `ContainerPlugin` with `library: {type: "systemjs"}` and served on any URL different from the host application's root can't have its exposed module loaded.

**If the current behavior is a bug, please provide the steps to reproduce.**

A host app served on `http://localhost:9000` loads the 'http://localhost:9001/remoteEntry.js' file using SystemJS : 
```js
SystemJS.import('http://localhost:9001').then(container => {
[...]
})
```
When trying to call `container.get('./someModule')`, the corresponding request is sent to `http://localhost:9000/some_module_js.js` instead of `http://localhost:9001/some_module_js.js`

Here is a repository to reproduce the problem : https://github.com/fbessou/webpack-systemjs-container
See `README.md` for instructions on how to reproduce.

**What is the expected behavior?**

The exposed modules should be loaded from a path relative to the remoteEntry.js file and on the same origin.

**Other relevant information:**
webpack version:  5.38.1
Node.js version: 14.17.0
Operating System: Debian 10 (buster)
Additional tools:

[a comment by an automated account is not shown]

**Comment by alexander-akait** [comment]
Why you use `ContainerPlugin`? You should not use it if you don't understand what you do. And you need other plugin on other side or it will not work.

**Reply by alexander-akait** [reply]
For server development you need specify host and port like here https://github.com/module-federation/module-federation-examples/blob/master/startup-code/app1/webpack.config.js#L38

**Reply by fbessou** [reply]
Thanks for your answer :)

> Why you use `ContainerPlugin`? You should not use it if you don't understand what you do.

I updated the example repo to replace the `ContainerPlugin` with the `ModuleFederationPlugin` to show that the problem is still there. I just wanted to show that problem was comming from the `ContainerPlugin`.



> And you need other plugin on other side or it will not work.
> For server development you need specify host and port like here https://github.com/module-federation/module-federation-examples/blob/master/startup-code/app1/webpack.config.js#L38

My use case is an host application with a plugin system. The plugins list is not known at compilation time so I can't use remotes.
My real host application's webpack configuration is similar to https://github.com/module-federation/module-federation-examples/blob/master/advanced-api/dynamic-remotes/app1/webpack.config.js . In the example repository the `ModuleFederationPlugin` is no longer useful on the host application since there is no shared dependency.

Note that my example is quite similar to the dynamic-remotes example made by @ScriptedAlchemy (https://github.com/module-federation/module-federation-examples/tree/master/advanced-api/dynamic-remotes). The only difference is that when using SystemJS to load the container (instead of a `<script>` tag) the publicPath is wrongly detected.

**Reply by alexander-akait** [reply]
> the publicPath is wrongly detected.

Yes, it is limitation, can you look at https://github.com/webpack/webpack/pull/10703#issuecomment-617822314?

**Reply by fbessou** [reply]
From what I remember, the proposed solution (merging a file into the entry) didn't work either. I will update the example repository and add more details here. :)

**Reply by fbessou** [reply]
I checked and the problem with this approach is that the `get` and `init` functions are erased and `setPublicPath` is the only export available on the container entry. I created a branch on the example repository to show this : https://github.com/fbessou/webpack-systemjs-container/tree/set-public-path

**Reply by alexander-akait** [reply]
You need to use plugin here https://github.com/fbessou/webpack-systemjs-container/tree/set-public-path too, otherwise webpack doesn't know how to load some modules

**Reply by fbessou** [reply]
I am not sure to understand what you mean. Should I also add the ModuleFederationPlugin to the host ?

**Reply by alexander-akait** [reply]
on host and remote, look at generated code, it is not hard to see how you can use it
